### PR TITLE
[REVIEW] NXCM-4979: Tasks dissapearing

### PR DESCRIPTION
**Pull request by cstamas**
Related issue
https://issues.sonatype.org/browse/NXCM-4979

This pull does not fix the issue, as it was not reproduced by us. Here are only fixes found during code reading:
- task config manager was leaving configuration "dirty" in some cases
- added a "diagnostic log" that can be enabled on certain systems, that should provide us more information about the main issue. Enabling is possible without putting Nexus into DEBUG mode, to end up with logs filling disk.
- reformat

Related changes in Task Scheduler:
- added logging, to not "swallow" exceptions https://github.com/sonatype/sisu-task-scheduler/commit/a647589427e25d25f9e4fd067456f45c38b4d51a
- variable name cleanup https://github.com/sonatype/sisu-task-scheduler/commit/119dd891bb7d560ee747c8e33fa71066f6f7f3d4
- added "reference" UT https://github.com/sonatype/sisu-task-scheduler/commit/119dd891bb7d560ee747c8e33fa71066f6f7f3d4

As discussed on issue, this change does not make things worse, only improves logging. Hence, is safe to be merged.

### CODE REVIEW

**Comment by kellyrob99** [comment]
Hard to tell what's actually changed here with all the code reformatting mixed in... All I really spotted was the system prop handling for debug messages. Could you please highlight anything pertinent that I am missing?

File: nexus-core/src/main/java/org/sonatype/scheduling/DefaultTaskConfigManager.java [hunk 1]
```diff
@@ -55,12 +56,15 @@
...
  */
-@Component ( role = TaskConfigManager.class )
+@Component( role = TaskConfigManager.class )
 public class DefaultTaskConfigManager
     extends AbstractConfigurable
     implements TaskConfigManager
 {
-    private final Logger logger = LoggerFactory.getLogger(  getClass() );
+    private final boolean NXCM4979DEBUG = SystemPropertiesHelper.getBoolean(
+        "org.sonatype.scheduling.TaskConfigManager.nxcm4979debug", false );
```
Comment (jdillon) [inline nexus-core/src/main/java/org/sonatype/scheduling/DefaultTaskConfigManager.java]: btw, this is really really really bad naming.  we should never be using jira issue ids in this fashion in our codebase... _ever_

you have have simply used trace logging here and enabled trace for this logger w/o needing this complexity at all.

File: nexus-core/src/main/java/org/sonatype/scheduling/DefaultTaskConfigManager.java [hunk 2]
```diff
@@ -204,20 +205,26 @@ void initializeTasks( Scheduler scheduler, List<CScheduledTask> tasks )
                 if ( foundTask != null )
                 {
                     tasks.remove( foundTask );
-                    
+
                     storeableTask.setLastRun( foundTask.getLastRun() );
                 }
 
                 tasks.add( storeableTask );
+            }
```
Comment (jdillon) [inline nexus-core/src/main/java/org/sonatype/scheduling/DefaultTaskConfigManager.java:209]: relevant non-logging code change here ^^^

File: nexus-core/src/main/java/org/sonatype/scheduling/DefaultTaskConfigManager.java [hunk 3]
```diff
@@ -233,15 +240,21 @@ void initializeTasks( Scheduler scheduler, List<CScheduledTask> tasks )
             if ( foundTask != null )
             {
                 tasks.remove( foundTask );
+            }
```
Comment (jdillon) [inline nexus-core/src/main/java/org/sonatype/scheduling/DefaultTaskConfigManager.java:239]: relevant non-logging code change here ^^^

File: nexus-core/src/main/java/org/sonatype/scheduling/DefaultTaskConfigManager.java [hunk 4]
```diff
@@ -251,42 +264,52 @@ void initializeTasks( Scheduler scheduler, List<CScheduledTask> tasks )
     public SchedulerTask<?> createTaskInstance( String taskType )
         throws IllegalArgumentException
     {
```
Comment (jdillon) [inline nexus-core/src/main/java/org/sonatype/scheduling/DefaultTaskConfigManager.java:262]: afacit vvv is all noise

File: nexus-core/src/main/java/org/sonatype/scheduling/DefaultTaskConfigManager.java (hunk 1 above)
Comment (peterlynch) [inline nexus-core/src/main/java/org/sonatype/scheduling/DefaultTaskConfigManager.java]: enabling with trace logging sounds fine with me would rather not tie the logging to a particular bug no.

File: nexus-core/src/main/java/org/sonatype/scheduling/DefaultTaskConfigManager.java [hunk 5]
```diff
@@ -204,20 +201,26 @@ void initializeTasks( Scheduler scheduler, List<CScheduledTask> tasks )
...
+            }
 
-                try
-                {
-                    applicationConfiguration.saveConfiguration();
-                }
-                catch ( IOException e )
-                {
-                    logger.warn( "Could not save task changes!", e );
-                }
```
Comment (cstamas) [inline nexus-core/src/main/java/org/sonatype/scheduling/DefaultTaskConfigManager.java:220]: Bellow added trace logging (is new)

File: nexus-core/src/main/java/org/sonatype/scheduling/DefaultTaskConfigManager.java [hunk 6]
```diff
@@ -204,20 +201,26 @@ void initializeTasks( Scheduler scheduler, List<CScheduledTask> tasks )
...
+            if ( logger.isTraceEnabled() )
+            {
+                logger.trace( "Task with ID={} added, config {} modified.", task.getId(), storeableTask != null ? "IS"
+                    : "is NOT", new Exception( "This is an exception only to provide caller backtrace" ) );
+            }
+
+            try
```
Comment (cstamas) [inline nexus-core/src/main/java/org/sonatype/scheduling/DefaultTaskConfigManager.java:217]: This is non-logging related change: in some cases config was left "in limbo": appConfig.save was not invoked, even if above config of this component was made "dirty": line 193 `List<CScheduledTask> tasks = getCurrentConfiguration( true );`

Now config is _always_ saved, as we _always_ make it dirty, even if sometimes no actual change happens. Leaving config dirty is wrong.

File: nexus-core/src/main/java/org/sonatype/scheduling/DefaultTaskConfigManager.java [hunk 7]
```diff
@@ -233,15 +236,21 @@ void initializeTasks( Scheduler scheduler, List<CScheduledTask> tasks )
...
 
-                try
-                {
-                    applicationConfiguration.saveConfiguration();
-                }
-                catch ( IOException e )
-                {
-                    logger.warn( "Could not save task changes!", e );
-                }
+            if ( logger.isTraceEnabled() )
```
Comment (cstamas) [inline nexus-core/src/main/java/org/sonatype/scheduling/DefaultTaskConfigManager.java:241]: Added trace logging (new)

File: nexus-core/src/main/java/org/sonatype/scheduling/DefaultTaskConfigManager.java [hunk 8]
```diff
@@ -233,15 +236,21 @@ void initializeTasks( Scheduler scheduler, List<CScheduledTask> tasks )
...
+            {
+                logger.trace( "Task with ID={} removed, config {} modified.", task.getId(), foundTask != null ? "IS"
+                    : "is NOT", new Exception( "This is an exception only to provide caller backtrace" ) );
+            }
+
+            try
```
Comment (cstamas) [inline nexus-core/src/main/java/org/sonatype/scheduling/DefaultTaskConfigManager.java:247]: Same reasoning as for addTask() change (config left in limbo)

**Pull request merged**